### PR TITLE
Add query params dialog

### PR DESCRIPTION
**Pull request by FranciscoDA**
This commit adds support for query parameters using GDA's placeholder syntax.

E.g:
```sql
SELECT * FROM myTable LIMIT ##myLimit::uint;
```

The user is prompted to input the placeholder values before running a query that contains GDA placeholders

Screenshot from 2020-04-11 19-44-57

### CODE REVIEW

**Comment by FranciscoDA** [comment]
I think the valac compiler from the CI pipeline doesn't support `int.try_parse` and `uint.try_parse`. A quick google search showed that support for it was only added 5 months ago:
https://gitlab.gnome.org/GNOME/vala/commit/ae7280960f2d4534f3ffd5668615d1e733002b98

I'm running valac version 0.48.3 and it's working on my machine.

We could change those calls to `int.parse` and `uint.parse` if we really need to maintain backwards compatibility, but we would lose type checking for inputted parameters.

File: src/Layouts/Views/Query.vala [hunk 1]
```diff
@@ -396,6 +387,35 @@ public class Sequeler.Layouts.Views.Query : Gtk.Grid {
     }
 
     public void run_query (string query) {
+        Gda.Statement statement;
+        Gda.Set params;
+        try {
+            statement = window.main.connection_manager.parse_sql_string (query, out params);
+        }
+        catch (GLib.Error ex) {
```
Comment (Alecaddd) [inline src/Layouts/Views/Query.vala:395]: Put the catch inline 
```suggestion
       } catch (GLib.Error ex) {
```

File: src/Widgets/QueryParamsDialog.vala [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        }
+
+        content.add (form_grid);
+
+        var run_button = new Sequeler.Widgets.RunQueryButton ();
+        run_button.set_sensitive (true);
+        add_action_widget (run_button, Action.RUN_QUERY);
+
+        // TODO: cancel_button should keep a consistent style
+        var cancel_button = new Sequeler.Widgets.CancelQueryButton ();
```
Comment (Alecaddd) [inline src/Widgets/QueryParamsDialog.vala:149]: Same thing here for the run query button.
No need to add a dedicated class if this is not repeated or reused.

File: src/meson.build [hunk 3]
```diff
@@ -47,6 +47,9 @@ executable(
     'Services/Types/SQLite.vala',
 
     'Widgets/ConnectionDialog.vala',
+	'Widgets/QueryParamsDialog.vala',
```
Comment (Alecaddd) [inline src/meson.build:50]: Fix the indentation here.

File: src/Layouts/Views/Query.vala [hunk 4]
```diff
@@ -292,16 +292,7 @@ public class Sequeler.Layouts.Views.Query : Gtk.Grid {
...
-        run_button.get_style_context ().add_class ("suggested-action");
-        run_button.get_style_context ().add_class ("notebook-temp-fix");
-        run_button.always_show_image = true;
-        run_button.image = new Gtk.Image.from_icon_name ("media-playback-start-symbolic", Gtk.IconSize.BUTTON);
-        run_button.image.valign = Gtk.Align.CENTER;
-        run_button.can_focus = false;
-        run_button.margin = 10;
-        run_button.sensitive = false;
-        run_button.tooltip_markup = Granite.markup_accel_tooltip ({"<Control>Return"}, _("Run Query"));
+        run_button = new Sequeler.Widgets.RunQueryButton ();
```
[the remark this answers is not in the document]
Comment (Alecaddd) [reply src/Layouts/Views/Query.vala:295]: Ah, I see from the screenshot that the run query button is reused, so indeed this is useful.
Be sure to respect the semantic order [Cancel] [Run] for the dialog.

**Comment by Alecaddd** [comment]
Thanks for the screenshot.
How does the dialog behave when handling multiple fields?
Let's say the user has a very complex query with 40 or 50 fields?

**Comment by FranciscoDA** [comment]
> Thanks for the screenshot.
> How does the dialog behave when handling multiple fields?
> Let's say the user has a very complex query with 40 or 50 fields?

Not very gracefully, I'm afraid

Screenshot from 2020-04-11 19-48-22

File: src/Widgets/QueryParamsDialog.vala (hunk 2 above)
Comment (FranciscoDA) [reply src/Widgets/QueryParamsDialog.vala]: Moved to `build_cancel_button`

File: src/Layouts/Views/Query.vala (hunk 4 above)
[the remark this answers is not in the document]
Comment (FranciscoDA) [reply src/Layouts/Views/Query.vala:295]: I've just updated the pull request and the screenshot to match that order :+1:

**Comment by Alecaddd** [comment]
> Not very gracefully, I'm afraid

LOL, that's neat :sweat_smile: 

Mhhh...I guess we could solve it by using a Gtk Grid with 2 columns, and wrapped the content around a ScrolledWindow, so in case the user has multiple fields, the dialog doesn't explode.

**Comment by FranciscoDA** [comment]
> > Not very gracefully, I'm afraid
> 
> LOL, that's neat sweat_smile
> 
> Mhhh...I guess we could solve it by using a Gtk Grid with 2 columns, and wrapped the content around a ScrolledWindow, so in case the user has multiple fields, the dialog doesn't explode.

Fixed :100: 
Screenshot from 2020-04-11 20-41-12

**Comment by Alecaddd** [comment]
Awesome, thank you so much.
There's a small merge conflict due to a recent PR I merged to fix some nasty segfaults when handling async and multithread methods.

**Comment by FranciscoDA** [comment]
Yup, I think moving a statement two lines below was enough to confuse git and make it mark something else as the conflict :monocle_face:

**Comment by Alecaddd** [comment]
Thanks.
I'm reviewing it locally and it works pretty well.
I will update a couple of things because the new dialog doesn't work super well with the elementary theme, nothing too drastic tho.

This is a really neat implementation, but I'm still struggling to understand the benefits of it.
What are the reasoning behind a user preferring to use placeholder syntax instead of writing those values directly in the query?

**Comment by Alecaddd** [comment]
Actually, now that I'm looking at it, I think this dialog will be extremely useful when I will implement the "double click to edit attribute".
I will merge this soon, thanks again.

**Comment by FranciscoDA** [comment]
Some queries require the same value to be used multiple times. Placeholders will help to enter that value faster and will be less error-prone, since there's no chance you missed a value that had to be replaced. Also allows you to mindlessly copy-paste a query and let the program ask you for the required parameters.

Long example ahead:

Suppose I have a table `trades` with a schema like so:
```sql
CREATE TABLE trades (
ask_user_id INT,
bid_user_id INT,
bid_amount DECIMAL(10,2),
ask_amount DECIMAL(10,2),
bid_currency VARCHAR,
ask_currency VARCHAR
);
```

Say that I want to retrieve all the operations of a user, I need to use union queries to fetch the data from that table, but I also need all the queries in the union to filter by the same user id, like so:
```sql
SELECT bid_currency, 'SOLD', -sum(bid_amount) FROM trades WHERE bid_user_id = ##userid::int
UNION SELECT ask_currency, 'BOUGHT', sum(ask_amount) FROM trades WHERE bid_user_id = ##userid::int;
UNION SELECT bid_currency, 'BOUGHT', sum(bid_amount) FROM trades WHERE ask_user_id = ##userid::int
UNION SELECT ask_currency, 'SOLD', -sum(ask_amount) FROM trades WHERE bid_user_id = ##userid::int
```
If I were to enter mismatching ids (as I have done a few times), the bought and sold amounts from the user would be wrong!

**Comment by Alecaddd** [comment]
That makes sense, thanks for the explanation.

So, I'm updating a bit your patch to implement License comments, remove some redundancy in the code (eg. buttons don't need to be set insensitive if we immediately update them to be sensitive on creation), the query run accelerator doesn't work on the dialog button, and writing an error message for the `store_parsed_value ()` method.

**Pull request merged**